Add books of public names/qualifiers/alias/titles/domains

a2line opened the pull request:
commits can probably be squashed

hgouraud committed:
add five books for mass renaming in databases, enhance book of sources

Wizards have new books for public_names, qualifiers, aliases, domains and titles.
The book of sources now includes all individual and familal sources.
The list of inviduals (m=L) for a specific entry supports all books and
allows to verify each targeted individual one by one before renamming by lot.

a2line committed:
deploy cached gzipped datalists on forms

Populate all datalists with a js function using files created by the cache_files tool.
Ctrl+left-clic on inputs allows to open current value in a new tab in associated book.
Fix the year regexp pattern in both forms, remove unused %nbs variable.

hgouraud committed:
move %base.name; variable from perso to templ.ml

hgouraud committed:
add the cache_files tool and its report table in gwsetup

Update gwsetup HTML and lexicon to support all cache_files parameters.
The table of results with name, file size and quantity of entries of the cache
files is now reported from the console log to gwsetup.

hgouraud committed:
rewrite cache_files for autocompletion datalists, add gzip compression support

* New datalist caches for the new books (titles, domains, public_names, qualifiers and aliases).
* New “-bd” option for cache_files tool.
* Remove all <option> HTML tags in the cache files to reduce their size by nearly 50% (A2).
* New Opam dependancy "camlzip" added to compress the cache files with gzip (A2).
* Format the “cache_file.exe” report log as a table (A2+hg).

ilankri commented:
> commits can probably be squashed

Please do not squash unrelated changes, thanks!

a2line committed:
new functions to improve shortcuts from forms to book usability

— The ctrl-click shortcut from inputs in forms goes directly to the edited entry
in books, facilitating mass renaming or fixing erroneous entries with around suggestions.
It handles subplaces specificities for the book of places.
— Add a ‘clear button’ (X) to all inputs to facilitate autocompletion: it allows users
to quickly empty the inputs and to see all options offered by the cached datalist.
It supports a new s2 parameter to change the book input from URLs.
— A generic function to put the cursor in inputs with an autofocus parameter at the end of the input.
— A last function highlights greek and cyrilic initial in books using regex.min.js.
This library could be later usefull in books to better identify script mixup, for example inputs that
are in latin and greek and that should not exist.

a2line committed:
remove random button if base is empty

michelmno commented:
hgouraud/domain2 to be rebased on top of  geneweb/master  that already merged two of the nine commits since PR #1855 merge.
cid 8ed25f5 test/testgalichet.gw updated concequence of PR #1840
cid 0d63afa test/run-GW-test.sh better grep output if error log

michelmno commented:
As follow up of cid https://github.com/geneweb/geneweb/pull/1858/commits/df581c022b5578cf707f217e1299e0d854a7f1b6 added in this PR,
could I suggest to change book title `Nom public` to `Noms publics` (and `Public name` to `Public names`
to have plurals as all other books titles ?
![Capture d’écran du 2024-08-16 17-46-19](https://github.com/user-attachments/assets/09de77e1-1b17-42d0-b360-a629c88f6c95)
![Capture d’écran du 2024-08-16 17-47-04](https://github.com/user-attachments/assets/216cc9af-f6a0-4106-9157-d3fb36e743ea)

michelmno commented:
Another question about same cid https://github.com/geneweb/geneweb/commit/df581c022b5578cf707f217e1299e0d854a7f1b6
There are new white triangles in new books head  buttons
like `Go to section 'A' ` in addition to  other part of the button that `Search 'A'`
`http://localhost:2317/basefusionnee?lang=en&m=MOD_DATA&data=qual#_A`
`http://localhost:2317/basefusionnee?lang=en&m=MOD_DATA&data=qual&s=A`
Why not to do the same for previous books pages ?
![Capture d’écran du 2024-08-16 18-50-01](https://github.com/user-attachments/assets/687f6288-60ac-491b-924c-475f70107b66)

michelmno commented:
About cid https://github.com/geneweb/geneweb/pull/1858/commits/2c1f06cbb5283152b69014cf11a3bccdb4fe12f5
Still have reported random error in books pages with empty base:
![Capture d’écran du 2024-08-16 19-25-24](https://github.com/user-attachments/assets/5a7cd1f5-abd9-4cc0-873f-21f18957a638)

a2line commented:
@michelmno:
Ok, welcome.txt has a generic function calling plural of books <s>and it was only a lexicon entry duplicate: I have merged them, it's fixed on this branch</s> I have left the duplicate lexicon entry and use it in template, but this should be cleaned one day or another.

The triangle arrows to navigate are not from this PR, they come from PR from last december already in master. Moreover I don't understand your request!

Edited: The fix for random button removes the button itself, if the number of individual in the database is null, but we see it on your screenshot?!

michelmno commented:
> Edited: The fix for random button removes the button itself, if the number of individual in the database is null, but we see it on your screenshot?!

I agree :) this is why I reported the problem here, still has problem with last update of hgouraud/domain2 (cid 05d6cdc73)
![Capture d’écran du 2024-08-21 18-32-41](https://github.com/user-attachments/assets/6a2f5110-01da-437c-abde-5803feba82d6)

michelmno commented:
> The triangle arrows to navigate are not from this PR, they come from PR from last december already in master. Moreover I don't understand your request!

Thank you,  my fault !
I did not saw that before because my DB had too many entries (greater than default book_max_results value of gwf) so the other book first pages  did not have those small triangles, I agree I am able to see them if digging in next pages.

hgouraud committed:
rename title to book_name in updateDataDisplay

hgouraud committed:
some better critical error messages

hgouraud committed:
rework wiki.ml to print original fn, sn

a2line committed:
remove empty bar when no password is set

hgouraud committed:
add a_of_b and url_encode/decode for welcome page

michelmno commented:
@a2line @hgouraud  are you solving the same problem in cid  93ca75e in this PR #1858 and cid 90b32de4c in other PR #1874  ?  I do not know how to reconcile them if want to merge two branches for test.

If interested I made a trial of merge of two branches for build test in https://github.com/michelmno/geneweb/commits/domains2_merged_on_reorg12/  
* there are two commits to solve fmt errors
* I marked few commits with [WIP]  when required reconcile between two branches.

a2line committed:
allow link to user.name and user.ident

add warning for those case with user.key
remove span

michelmno commented:
I identified two new failures in this branch, with help of test/run-GW-test.sh
The first one
```
curl http://localhost/cgi-bin/gwtst?b=basefusionnee&w=azerty:dcba&
  var errors_list = "\u{000A}";
/tmp/gwd_cgi_tst_error.log reported traces (empty if no failure):
File ./etc/welcome.txt, line 105, characters 17-37:
Failed - int value expected
--
File ./etc/welcome.txt, line 106, characters 15-35:
Failed - int value expected
at least 2 detected error(s).
```
With help of git bisect I identified the failing commit:
```
git bisect start 
git bisect run  ...
...
c2b98cf43037d0e768e8d6e951f01fc0678851c5 is the first bad commit
commit c2b98cf43037d0e768e8d6e951f01fc0678851c5
Author: A2 <a2line@gmail.com>
Date:   Sat Aug 24 03:47:28 2024 +0200

    fixup! remove random button if base is empty
```

michelmno commented:
The second failure is `Unbound variable(s): %opt?`
detected with an update of test/run-GW-test.sh (see PR #1890)
and using git bisect I identified cid aa3ff7b as source of the failure
```
curl http://localhost/cgi-bin/gwtst?b=basefusionnee&w=azerty:dcba&m=L&data=place;&bi=on&ba=on&de=on&bu=on&ma=on&k=Australie&nb=1&i0=1197&p0=Australie
  var errors_list = "\u{000A}Unbound variable(s): %opt?. ";
/tmp/gwd_cgi_tst_error.log reported traces (empty if no failure):
at least 1 detected error(s).
aa3ff7b107985146caabb9c0067ea84538e04ed7 is the first bad commit
commit aa3ff7b107985146caabb9c0067ea84538e04ed7
Author: Henri83 <henri.gouraud@laposte.net>
Date:   Sun Jun 23 14:39:55 2024 +0200

    add five books for mass renaming in databases, enhance book of sources
```

a2line committed:
new minimal generation v0 parameter for tables

* The new parameter allows users to print tables from a specific range of generations.
The new evar v0 represents the starting generation.
* Update all titles of tables to reflect this new feature.
* Add a new short form in the toolbar containing v0 (min) and v (max). Previously the maximummodifiable
could only be modified using the ±1 button (one by one).
* Improve highligh/spouses info buttons accessibility with clearer translations of their states.
* Improve readability of table using lighten border. Yellow and gray highligh are also lighter.
* Fix various button bugs switching tables.

a2line committed:
fix %s variable in “having as children” lexicon entry

fix #1906

hgouraud committed:
escape ' in html tag attributes

hgouraud committed:
better witness handling in timeline

a2line commented:
Push forced mainly to fix a rare case where an user (hg!) could have a personnal lexicon translation for `[list]` and the jquery selector to populate the datalists on the fly in forms was failing. We have to be carefull when using brackets for javascript in .txt template files: they need to be escaped with `%` overwise the tranlastion is called if present (so here changed to `%[list%]`). The ctrl-clic function is now also autonomous and does not depend on datalist presence anymore.

hgouraud committed:
suppress -opt in setup.ml, print to stderr in cache_file

hgouraud committed:
correctly handle & character in URLs using uri_encode

hgouraud committed:
better support of ampersand & character in filenames of images

a2line committed:
more scripts already used on Roglo

michelmno commented:
To avoid build error as reported by last checks actions (https://github.com/geneweb/geneweb/actions/runs/11685476116/job/32539004109?pr=1858), should cherry pick commit https://github.com/michelmno/geneweb/commit/1799ec428786ef3a0d8bec3c8dbf77e072205e30
Furthermore, this hgouraud:domains2 branch should be rebased on top of geneweb:master that has at least correction of issue #1854
